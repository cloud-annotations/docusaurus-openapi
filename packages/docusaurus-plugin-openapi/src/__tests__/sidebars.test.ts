/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import loadSidebars from '../sidebars';

/* eslint-disable global-require, import/no-dynamic-require */

describe('loadSidebars', () => {
  const fixtureDir = path.join(__dirname, '__fixtures__', 'sidebars');
  test('sidebars with known sidebar item type', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars.json');
    const result = loadSidebars([sidebarPath]);
    expect(result).toMatchSnapshot();
  });

  test('sidebars with deep level of category', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-category.js');
    const result = loadSidebars([sidebarPath]);
    expect(result).toMatchSnapshot();
  });

  test('sidebars shortand and longform lead to exact same sidebar', async () => {
    const sidebarPath1 = path.join(fixtureDir, 'sidebars-category.js');
    const sidebarPath2 = path.join(
      fixtureDir,
      'sidebars-category-shorthand.js',
    );
    const sidebar1 = loadSidebars([sidebarPath1]);
    const sidebar2 = loadSidebars([sidebarPath2]);
    expect(sidebar1).toEqual(sidebar2);
  });

  test('sidebars with category but category.items is not an array', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-category-wrong-items.json',
    );
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"category\\",\\"label\\":\\"Category Label\\",\\"items\\":\\"doc1\\"}. \\"items\\" must be an array."`,
    );
  });

  test('sidebars with category but category label is not a string', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-category-wrong-label.json',
    );
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"category\\",\\"label\\":true,\\"items\\":[\\"doc1\\"]}. \\"label\\" must be a string."`,
    );
  });

  test('sidebars item doc but id is not a string', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-doc-id-not-string.json',
    );
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"doc\\",\\"id\\":[\\"doc1\\"]}. \\"id\\" must be a string."`,
    );
  });

  test('sidebars with first level not a category', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-first-level-not-category.js',
    );
    const result = loadSidebars([sidebarPath]);
    expect(result).toMatchSnapshot();
  });

  test('sidebars link', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link.json');
    const result = loadSidebars([sidebarPath]);
    expect(result).toMatchSnapshot();
  });

  test('sidebars link wrong label', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link-wrong-label.json');
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"link\\",\\"label\\":false,\\"href\\":\\"https://github.com\\"}. \\"label\\" must be a string."`,
    );
  });

  test('sidebars link wrong href', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link-wrong-href.json');
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"link\\",\\"label\\":\\"GitHub\\",\\"href\\":[\\"example.com\\"]}. \\"href\\" must be a string."`,
    );
  });

  test('sidebars with unknown sidebar item type', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-unknown-type.json');
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Unknown sidebar item type [superman]. Sidebar item={\\"type\\":\\"superman\\"} "`,
    );
  });

  test('sidebars with known sidebar item type but wrong field', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-wrong-field.json');
    expect(() =>
      loadSidebars([sidebarPath]),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Unknown sidebar item keys: href. Item: {\\"type\\":\\"category\\",\\"label\\":\\"category\\",\\"href\\":\\"https://github.com\\"}"`,
    );
  });

  test('no sidebars', () => {
    const result = loadSidebars(null);
    expect(result).toEqual({});
  });

  test('sidebars with category.collapsed property', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-collapsed.json');
    const result = loadSidebars([sidebarPath]);
    expect(result).toMatchSnapshot();
  });

  test('sidebars with category.collapsed property at first level', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-collapsed-first-level.json',
    );
    const result = loadSidebars([sidebarPath]);
    expect(result).toMatchSnapshot();
  });
});
