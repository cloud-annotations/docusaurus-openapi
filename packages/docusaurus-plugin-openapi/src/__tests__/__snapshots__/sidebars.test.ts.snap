// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadSidebars sidebars link 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "href": "https://github.com",
          "label": "category",
          "type": "link",
        },
      ],
      "label": "Test",
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with category.collapsed property 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "collapsed": false,
          "items": Array [
            Object {
              "id": "doc1",
              "type": "doc",
            },
          ],
          "label": "Introduction",
          "type": "category",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "collapsed": false,
          "items": Array [
            Object {
              "id": "doc2",
              "type": "doc",
            },
          ],
          "label": "Powering MDX",
          "type": "category",
        },
      ],
      "label": "Reference",
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with category.collapsed property at first level 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": false,
      "items": Array [
        Object {
          "id": "doc1",
          "type": "doc",
        },
      ],
      "label": "Introduction",
      "type": "category",
    },
    Object {
      "collapsed": false,
      "items": Array [
        Object {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Powering MDX",
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with deep level of category 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "a",
          "type": "doc",
        },
        Object {
          "collapsed": true,
          "items": Array [
            Object {
              "collapsed": true,
              "items": Array [
                Object {
                  "id": "c",
                  "type": "doc",
                },
                Object {
                  "collapsed": true,
                  "items": Array [
                    Object {
                      "id": "d",
                      "type": "doc",
                    },
                    Object {
                      "collapsed": true,
                      "items": Array [
                        Object {
                          "id": "e",
                          "type": "doc",
                        },
                      ],
                      "label": "deeper more more",
                      "type": "category",
                    },
                  ],
                  "label": "level 4",
                  "type": "category",
                },
              ],
              "label": "level 3",
              "type": "category",
            },
            Object {
              "id": "f",
              "type": "doc",
            },
          ],
          "label": "level 2",
          "type": "category",
        },
      ],
      "label": "level 1",
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with first level not a category 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "greeting",
          "type": "doc",
        },
      ],
      "label": "Getting Started",
      "type": "category",
    },
    Object {
      "id": "api",
      "type": "doc",
    },
  ],
}
`;

exports[`loadSidebars sidebars with known sidebar item type 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "foo/bar",
          "type": "doc",
        },
        Object {
          "id": "foo/baz",
          "type": "doc",
        },
        Object {
          "href": "https://github.com",
          "label": "Github",
          "type": "link",
        },
        Object {
          "id": "hello",
          "type": "ref",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "hello",
          "type": "doc",
        },
      ],
      "label": "Guides",
      "type": "category",
    },
  ],
}
`;
